### ISSUE TITLE
message resend per unique topic

### ISSUE THREAD

**Issue by muglar** [bartbutenaers/node-red-contrib-msg-resend#2]
Have all the same functionality, however, duplicate such that there is a individual resend timer (set globally) for every unique topic and those message payloads would be resent on a case by case basis.

**Comment by bartbutenaers** [comment]
Hi Muglar,

At first sight it seemed to be a good idea, so I started coding this evening. 

Currently (version 0.0.5) displays following status, as soon as all messages have been resend:
![image](https://cloud.githubusercontent.com/assets/14224149/23827141/f7cfb7e4-06ac-11e7-9258-3603304a9292.png)
I could display that status when ALL messages of ALL topics have been send.
So far so good ...

Moreover currently (version 0.0.5) each time a message has been resend, the status displays the message counter:
![image](https://cloud.githubusercontent.com/assets/14224149/23827168/7292d81c-06ad-11e7-9779-7a72328c1027.png)
However this status becomes very **confusing** now. 
 
For example, suppose messages arrive for two different topics:
MSG1 _(topic1)_ -> MSG2 _(topic1)_ -> MSG3 _(topic1)_ -> MSG4 _(topic2)_ -> MSG5 _(topic1)_ -> MSG6 _(topic2)_ ...
The counter goes up and down all the time in the flow editor:
1 -> 2 -> 3 -> 1 -> 4 -> 2 -> ...

Any suggestions on how you had this in mind? 

Kind regards,
Bart

**Comment by muglar** [comment]
Hi Bart,

Thanks for the reply.  Your message resend feature is a great addition to
node red.  I'm going to be using it right after a RBE node.  For my use
case I would have the resend maximum at 0, so the the maximum reached
doesn't really matter.  In fact the status of the node really is that
import to me - similar to the rbe that handles multiple unique topics, this
is probably why the rbe does not show its status.

You could do the following.....not as good as you currently have for one
topic, but can't think of anything better.

X unique, Y re-sending, Z maximum reached.

**Comment by bartbutenaers** [comment]
Hey Muglar,

I understand that in your case the status texts are not that important.  But other use cases might require it, and I have to keep my node as general as possible.  

Indeed I could show **more general status texts** (like '_re-sending_' instead of '_sended 2x_') in case of topic-based resending.    It would have been nicer if the topic name itself could be added as prefix in the status text (like '_topic1 sended 2x_'), but that would become too long for most topics.

I will create a test version in the next days, so you can check whether it does the job.
Will keep you informed!

Bart

**Comment by bartbutenaers** [comment]
Hi Muglar,

Had to change more coding as I thought.  A new version is available on Github.  You can uninstall your current msg-resend node and install the test version directly from Github:
`npm install https://github.com/bartbutenaers/node-red-contrib-msg-resend.git`

Can you test whether it fits your needs?  When working correctly, I will publish the new (0.0.6) version to NPM for other Node-Red users.  The explanation is updated in the readme.md file.

I have tested it with this flow:

![image](https://cloud.githubusercontent.com/assets/14224149/23836404/d6aaefcc-0777-11e7-9ca6-e854b10aca10.png)

```
[{"id":"f5f9f981.d454f8","type":"msg-resend","z":"47b91ceb.38a754","interval":"1","maximum":"5","bytopic":false,"clone":false,"name":"TopicResending","x":547.2812042236328,"y":486.15631103515625,"wires":[["f2b5eef.f07871"]]},{"id":"7c11dd44.70fa94","type":"inject","z":"47b91ceb.38a754","name":"","topic":"topic1","payload":"FromTopic1","payloadType":"str","repeat":"","crontab":"","once":false,"x":176.27430725097656,"y":472.9583435058594,"wires":[["f5f9f981.d454f8"]]},{"id":"f2b5eef.f07871","type":"debug","z":"47b91ceb.38a754","name":"Resend test","active":true,"console":"false","complete":"payload","x":732.2812042236328,"y":486.0694885253906,"wires":[]},{"id":"f97ea211.76e1c","type":"inject","z":"47b91ceb.38a754","name":"","topic":"topic2","payload":"FromTopic2","payloadType":"str","repeat":"","crontab":"","once":false,"x":174.8958282470703,"y":521.8889465332031,"wires":[["f5f9f981.d454f8"]]},{"id":"32ca8e94.5ed3d2","type":"inject","z":"47b91ceb.38a754","name":"","topic":"","payload":"Empty topic","payloadType":"str","repeat":"","crontab":"","once":false,"x":153.8958282470703,"y":570.888916015625,"wires":[["f5f9f981.d454f8"]]},{"id":"660a9e9c.5c0e3","type":"inject","z":"47b91ceb.38a754","name":"Set resend_by_topic","topic":"","payload":"","payloadType":"date","repeat":"","crontab":"","once":false,"x":173.8958282470703,"y":624.888916015625,"wires":[["7ef2a8f8.31a7d8"]]},{"id":"7ef2a8f8.31a7d8","type":"change","z":"47b91ceb.38a754","name":"","rules":[{"t":"set","p":"resend_by_topic","pt":"msg","to":"true","tot":"bool"},{"t":"set","p":"resend_ignore","pt":"msg","to":"true","tot":"bool"}],"action":"","property":"","from":"","to":"","reg":false,"x":375.82640075683594,"y":625.1458435058594,"wires":[["f5f9f981.d454f8"]]}]
```

P.S. I have added some kind of progress bar (using dots) in the node status text **'resending ...'**, so you can see visually that the node is still working for endless resending (as in your case).

Greetings,
Bart

**Comment by muglar** [comment]
Hi Bart,

Installed and tested.  Works great if you don't redeploy or make any
changes on the fly! Perfect!

Abnormal behavior found.

1.) If topic dependent is chosen and a message received into the node
throws the following error (see screen pic below) (this doesn't seem to
affect its operation)  "msg : error "ReferenceError: timer is not defined".

*2.) If you change any of the settings after initial configuration and a
message has already been received and then redeploy the node stills works
on the original settings as well as the new settings, as such multiple
messages are sent for the same unique topic at different rates.  You may
have not seen this because you had the max count set at 5 and 1 second
internal.  For a resent max count of zero, this causes abnormal behavior.
Ideally the node would apply the new global settings to the node and dump
all the existing timers.  This may be difficult, so maybe just completely
flush the node and start fresh on all redeploys and wait for the messages
to flow in?*
3.) High input rate shown when receiving greater than 3 msg/sec.  I'm going
to be sending 300 msg/sec.....do you think this is going to be a problem?


[image: Inline image 2]

**Comment by muglar** [comment]
One other weird thing

If you completely delete the node it somehow still sends messages.  How is
that even possible?!

[image: Inline image 1]

**Comment by bartbutenaers** [comment]
Hey Muglar,

I'm back ...  Thanks for your extensive feedback!  You are promoted as my private tester from now on ;-)

I have submitted **two fixes** on Github:

- I removed an obsolete line, causing the "timer is not defined" error.  The annoying error is gone, but it doesn't have any impact on the rest of the node (as you already mentioned).

- I fixed the code that is responsible for cleaning up the timers, at the 'close' event (e.g. at deploy, redeploy, node removal, ...).  Seemed that a bug in my cleanup code, caused Node-Red to call my cleanup code only once.  Very weird behaviour. 

If you could test it again, I hope all problems are solved.

The only thing that is not clear to me is your third question: 

> High input rate shown when receiving greater than 3 msg/sec.  I'm going to be sending 300 msg/sec.....do you think this is going to be a problem?

The purpose of the 'high input rate' is this: suppose you set a resend interval of 1 second.  When a message (with topic_1) arrives, a timer will be setup that resends this message every second.  However suppose a new message arrives (also topic_1) faster than 1 second: a new timer should be created and the previous timer (_that didn't get a chance to send anything!_) need to be removed.  _**It is useles to create timers that don't get a chance to do anything.  So a 'high input rate' error is displayed...**_

However, you are talking about 300 messages per second.  I assume these messages don't have the same topic?  If they have a different topic, there is no problem: the resend node will create a timer for each message, even if they arrive at higher rates.

I have created a test flow.  If you press the start button once, the function node will start to send (every 50 millisecond) a message (_each with a unique topic_) to the resend node:

![image](https://cloud.githubusercontent.com/assets/14224149/23875937/812850b6-083b-11e7-97cb-49fba648396f.png)

```
[{"id":"afd5e1ef.55ed5","type":"function","z":"6beebf75.ed0b5","name":"Msg topic factory","func":"// Repeat the msg every 50 milliseconds\nvar repeatInterval = 50;\n\nvar interval = setInterval(function() {\n    var counter = context.get('counter') || 0;\n    counter = counter + 1;\n    context.set('counter', counter);\n    \n    if(counter == 300) {\n        clearInterval(interval);\n    }\n    \n\tnode.send({topic: 'mytopic_' + counter});\n}, repeatInterval); \n\nreturn null;","outputs":1,"noerr":0,"x":523.8957672119141,"y":2328.666748046875,"wires":[["8aadda79.62e148"]]},{"id":"590c58f7.a29548","type":"inject","z":"6beebf75.ed0b5","name":"","topic":"","payload":"Start","payloadType":"str","repeat":"","crontab":"","once":false,"x":325.8957977294922,"y":2328.666748046875,"wires":[["afd5e1ef.55ed5"]]},{"id":"5a6b8368.eabd8c","type":"debug","z":"6beebf75.ed0b5","name":"Resend test","active":true,"console":"false","complete":"topic","x":924.8957977294922,"y":2328.666748046875,"wires":[]},{"id":"8aadda79.62e148","type":"msg-resend","z":"6beebf75.ed0b5","interval":"1","maximum":"3","bytopic":true,"clone":false,"name":"","x":739.8957977294922,"y":2328.666748046875,"wires":[["5a6b8368.eabd8c"]]}]
```

It seems to work fine: the resend node repeats every message 3 times. The only disadvantage I see is that the resend-node status text cannot be displayed correctly anymore at higher speeds.

Please let me know what's on your mind.

Bart

**Comment by muglar** [comment]
Glad I could help :).  Worked like a charm.  Liked your topic factory
example.  No issues and as stated before great addition to the flow library.

I understand the high input rate message now - thanks for the explanation.

Cheers,

Jason

**Comment by bartbutenaers** [comment]
Muglar,

I have published version 0.0.6 on NPM, so you can download it from there now.  I made one extra change tonight: the node status is not updated more than once a second (since the flow editor cannot display it anyway at such high rate).  Perhaps useful if you send lots of messages every second.

See you in the next issue ;-)
Bart
